gcputil/token_exchange.go

```diff
@@ -0,0 +1,152 @@
+
+	resp, err := client.Do(req)
+
+	if err != nil {
+		return nil, fmt.Errorf("sts/google: invalid response from Secure Token Server: %v", err)
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	if c := resp.StatusCode; c < 200 || c > 299 {
+		return nil, fmt.Errorf("sts/google: status code %d: %s", c, body)
+	}
+	var stsResp STSTokenResponse
+	err = json.Unmarshal(body, &stsResp)
+	if err != nil {
+		return nil, fmt.Errorf("sts/google: failed to unmarshal response body from Secure Token Server: %v", err)
+
+	}
+
+	return &stsResp, nil
+}
+
+func makeIAMRequest(ctx context.Context, endpoint string, r *IAMTokenExchangeRequest) (*IAMTokenResponse, error) {
+	c := cleanhttp.DefaultClient()
+	data := map[string]interface{}{
+		"scope": r.Scope,
+	}
+	b, err := json.Marshal(data)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest("POST", endpoint, bytes.NewBuffer(b))
+	if err != nil {
+		return nil, fmt.Errorf("iamCredentials/google: failed to properly build http request: %v", err)
+	}
+	req = req.WithContext(ctx)
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", r.STSAccessToken))
+	resp, err := c.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("iamCredentials/google: invalid response from IAM Credentials Server: %v", err)
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	if c := resp.StatusCode; c < 200 || c > 299 {
+		return nil, fmt.Errorf("iamCredentials/google: status code %d: %s", c, body)
+	}
+
+	var stsResp IAMTokenResponse
+	err = json.Unmarshal(body, &stsResp)
+	if err != nil {
+		return nil, fmt.Errorf("iamCredentials/google: failed to unmarshal response body from IAM Credentials Server: %v", err)
+
+	}
+
+	return &stsResp, nil
+}
+
+// STSTokenExchangeRequest contains fields necessary to make an STS token exchange.
+type STSTokenExchangeRequest struct {
+	GrantType          string
+	Audience           string
+	Scope              []string
+	RequestedTokenType string
+	SubjectToken       string
+	SubjectTokenType   string
+}
+
+// STSTokenResponse is used to decode the remote server response during an STS token exchange.
+type STSTokenResponse struct {
+	AccessToken     string `json:"access_token"`
+	IssuedTokenType string `json:"issued_token_type"`
+	TokenType       string `json:"token_type"`
+	ExpiresIn       int    `json:"expires_in"`
+	Scope           string `json:"scope"`
+	RefreshToken    string `json:"refresh_token"`
+}
+
+// IAMTokenExchangeRequest contains fields necessary to make an IAM token exchange.
+type IAMTokenExchangeRequest struct {
+	Scope          []string
+	STSAccessToken string
+}
+
+// IAMTokenResponse is used to decode the remote server response during an IAM token exchange.
+type IAMTokenResponse struct {
+	AccessToken string `json:"accessToken"`
```

Comment:
Should we leave a comment noting the difference between this and `access_token ` in the STSTokenResponse so it is more obvious?